User and Group Logging Updates

aembler opened the pull request:
This is ready to review.
It contains:
- [x] Logging users that enter groups
- [x] Logging users that exit groups
- [x] Logging add user
- [x] Logging update user
- [x] Logging change user password
- [x] Logging reset user password
- [x] Delete user
- [x] Add Group
- [x] Edit Group
- [x] Delete Group
- [x] Activate User
- [x] Deactivate user

KorvinSzanto commented:
When I add a user as admin I get:
```
users.INFO: User foo (ID 2) was added by code or an automated process. {"user_id":"2","user_name":"foo","operation":"add_user"} {"_SERVER":{"FCGI_ROLE":"RESPONDER","REMOTE_ADDR":"127.0.0.1"}}
```
(My log client screws up the json)

Not sure why exactly after looking at the implementation

KorvinSzanto commented:
I'm not sure if this is something busted in my setup, but if I enable user deactivation after 2 failures in 1000 seconds then fail twice I get:

```An exception occurred while executing 'INSERT INTO UserDeactivatedNotifications (nID, userID, actorID) VALUES (?, ?, ?)' with params [1, 2, null]:```

KorvinSzanto commented:
> @KorvinSzanto Is that JSON really screwed up though? It looks like it just has one column for the context and one for the extra params, which is what your server processor is adding.

You are totally right, I was thinking my processor was adding that info to the *context* and not to that extra param and my log client was just stripping top level commas. Heh.

> Do you need this method here for any reason?

I'm sure we can get rid of that detach, the only issue with that could be that the instance is cached in both this class method's memoized output and in the doctrine cache so doctrine will constinue to analyze it every time `->flush` is run. That's not that big of a deal under normal circumstances though.

I think it'd also work to cascade detach to the UserSignup instance, but I'm not sure. 

> I find that detach causes 10 problems for every one that it solves.

I agree, but only because I think we're misusing the ORM at this point by keeping so many instances attached to the entity manager. It's probably the biggest hurdle in getting a repeatable runtime working because it's so leaky in terms of memory and CPU.

The easiest way around that though is probably not to just start using detach in one spot, it's probably just to make it so that we can use `$em->clear()` when we need to. At any rate it deserves some more thought.

aembler commented:
@KorvinSzanto Gotcha. That makes sense. I just wanted to make sure that we weren't using detach for a specific reason here; I agree being able to use clear throughout the code to handle this would make things more performant and would be a good goal to work toward. 

Ok, this pul request is updated. Can you review it again? There are a quite a few changes since yesterday, including:

1. Removal of `detach` to fix an issue where UserSignup was being complained about when a login failed and a user was deactivated.
2. Fixing that user deactivation exception when actorID was null.
3. Added some functionality to advanced logging where you can specify Channels::META_CHANNEL_ALL and get an advanced configuration attached to all logging channels (so you don't have to specify all core channels in order to add our new fancy processor, for example.)
4. Added a new logging test for this use case.
5. Fixed a bug in password reset – it was firing password reset flow even if you got the password wrong (as long as the user's password was actually reset.)

KorvinSzanto commented:
:+1: Looks good to me

mlocati commented:
In order to make translators life easier, it'd be much better to use positional arguments (the order of the words can vary depending on the language).
So, instead of
`User %s (ID %s) created group %s (ID %s).`
We should have
`User %1$s (ID %2$s) created group %3$s (ID %4$s).`

aembler commented:
That makes sense – would you consider opening a pull request?

mlocati commented:
Let's just wait for complains by other translators to see if these strings should be changed 😆 ...
Mine was just a note for future changes :wink:

aembler commented:
Hah 👍
